### PR TITLE
Implement metadata interface for baremetal operator

### PR DESCRIPTION
**Pull request by dukov**
Metadata is required for cloud-init config drive which is used by
Ironic. User should be able to create custom metadata in a secret and
reference it in a BareMetalHost object.

### CODE REVIEW

**Comment by dukov** [comment]
#447

**Comment by dukov** [comment]
#447

File: pkg/controller/baremetalhost/host_config_data.go [hunk 1]
```diff
@@ -79,3 +79,16 @@ func (hcd *hostConfigData) NetworkData() (string, error) {
 		"networkData",
 	)
 }
+
+// MetaData get host metatdata
+func (hcd *hostConfigData) MetaData() (string, error) {
+	if hcd.host.Spec.MetaData == nil {
+		hcd.log.Info("MetaData is not set returning epmty(nil) data")
```
Comment (dhellmann) [inline pkg/controller/baremetalhost/host_config_data.go]:
```suggestion
		hcd.log.Info("MetaData is not set returning empty(nil) data")
```

File: pkg/controller/baremetalhost/host_config_data.go [hunk 2]
```diff
@@ -79,3 +79,16 @@ func (hcd *hostConfigData) NetworkData() (string, error) {
...
+	if hcd.host.Spec.MetaData == nil {
+		hcd.log.Info("MetaData is not set returning empty(nil) data")
+		return "", nil
+	}
+	return hcd.getSecretData(
+		hcd.host.Spec.MetaData.Name,
+		hcd.host.Spec.MetaData.Namespace,
+		"metaData",
+	)
+}
```
Comment (maelk) [inline pkg/controller/baremetalhost/host_config_data.go]: This should have tests like for networkData and userData

File: pkg/provisioner/ironic/ironic.go [hunk 3]
```diff
@@ -978,32 +978,45 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
...
+		metaData := map[string]interface{}{
+			"uuid":             string(p.host.ObjectMeta.UID),
+			"metal3-namespace": p.host.ObjectMeta.Namespace,
+			"metal3-name":      p.host.ObjectMeta.Name,
+			"local-hostname":   p.host.ObjectMeta.Name,
+			"local_hostname":   p.host.ObjectMeta.Name,
+		}
+		metaDataRaw, err := hostConf.MetaData()
+		if err != nil {
+			return result, errors.Wrap(err, "could not retrieve network data")
```
Comment (maelk) [inline pkg/provisioner/ironic/ironic.go]: metadata

File: pkg/provisioner/ironic/ironic.go [hunk 4]
```diff
@@ -978,32 +978,45 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
...
+		}
+		if metaDataRaw != "" {
+			if err = yaml.Unmarshal([]byte(metaDataRaw), &metaData); err != nil {
+				return result, errors.Wrap(err, "failed to unmarshal meta_data.json from secret")
```
Comment (maelk) [inline pkg/provisioner/ironic/ironic.go]: unmarshal metadata

**Comment by dukov** [comment]
> I am putting together a design doc on how to handle metadata with CAPM3. I am including this there, so there probably is no need to create a dedicated design doc for it. I'll link the PR.

Ok. Thank you.

**Comment by maelk** [comment]
@dhellmann Can we merge this ? The design doc is waiting for review. This change is straightforward and we need it as soon as possible.

**Comment by maelk** [comment]
This is the design doc btw https://github.com/metal3-io/metal3-docs/pull/79

**Comment by maelk** [comment]
The design proposal was accepted and merged. We should now proceed with merging this PR. @dukov could you please fix the conflict and we'll merge this. Thank you

**Comment by dukov** [comment]
> The design proposal was accepted and merged. We should now proceed with merging this PR. @dukov could you please fix the conflict and we'll merge this. Thank you

Done

**Pull request merged**